java-components/domain-proxy/common/src/main/java/com/redhat/hacbs/domainproxy/common/CommonIOUtil.java

```diff
@@ -1,84 +1,153 @@
 package com.redhat.hacbs.domainproxy.common;
 
 import java.io.IOException;
-import java.net.Socket;
-import java.net.SocketException;
 import java.nio.ByteBuffer;
-import java.nio.channels.AsynchronousCloseException;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
 import java.nio.channels.SocketChannel;
+import java.util.Iterator;
 
 import org.jboss.logging.Logger;
 
 public final class CommonIOUtil {
 
     private static final Logger LOG = Logger.getLogger(CommonIOUtil.class);
 
-    public static Runnable createSocketToChannelWriter(final int byteBufferSize, final Socket socket,
-            final SocketChannel channel) {
-        // Write from socket to channel
+    public static final String LOCALHOST = "localhost";
+    public static final int TIMEOUT_MS = 1000;
+
+    private CommonIOUtil() {
+    }
+
+    private enum Operation {
+        READ("Read"),
+        WRITE("Wrote");
+
+        final String descriptor;
+
+        Operation(final String descriptor) {
+            this.descriptor = descriptor;
+        }
+    }
+
+    public static Runnable createChannelToChannelBiDirectionalHandler(final int byteBufferSize,
+            final SocketChannel leftChannel,
+            final SocketChannel rightChannel) {
         return () -> {
-            int r;
-            final byte[] buf = new byte[byteBufferSize];
-            int bytesWritten = 0;
-            LOG.info("Writing from socket to channel");
-            try {
-                while ((r = socket.getInputStream().read(buf)) > 0) {
-                    channel.write(ByteBuffer.wrap(buf, 0, r));
-                    bytesWritten += r;
+            Thread.currentThread().setName("ChannelToChannelBiDirectionalHandler");
+            LOG.info("Connections opened");
+            int bytesReadLeft = 0;
+            int bytesReadRight = 0;
+            int bytesWrittenLeft = 0;
+            int bytesWrittenRight = 0;
+            final ByteBuffer buffer = ByteBuffer.allocate(byteBufferSize);
+            try (final Selector selector = Selector.open()) {
+                leftChannel.configureBlocking(false);
+                rightChannel.configureBlocking(false);
+                leftChannel.register(selector, SelectionKey.OP_READ);
+                rightChannel.register(selector, SelectionKey.OP_WRITE);
+                long bytesTransferredTime = System.currentTimeMillis();
+                while (leftChannel.isOpen() && rightChannel.isOpen()) {
```

Comment:
This while loop exits on time-out or closed connections, but the connections are closed only after this loop ends?

Reply:
That's right. We return early upon timeout, which causes the channels to be closed. We also close the channels explicitly if one of them closes (ending while loop iteration) - one circumstance where this might happen is if a client (i.e. maven or curl) terminates early.